src/latbin/lattice-lmrescore-kaldi-rnnlm.cc

```diff
@@ -0,0 +1,151 @@
+
+
+#include "base/kaldi-common.h"
+#include "fstext/fstext-lib.h"
+#include "lat/kaldi-lattice.h"
+#include "lat/lattice-functions.h"
+#include "rnnlm/rnnlm-lattice-rescoring.h"
+#include "util/common-utils.h"
+#include "nnet3/nnet-utils.h"
+
+int main(int argc, char *argv[]) {
+  try {
+    using namespace kaldi;
+    typedef kaldi::int32 int32;
+    typedef kaldi::int64 int64;
+
+    const char *usage =
+        "Rescores lattice with kaldi-rnnlm. This script is called from \n"
+        "scripts/rnnlm/lmrescore_rnnlm_lat.sh. An example for rescoring \n"
+        "lattices is at egs/swbd/s5/local/rnnlm/run_rescoring.sh \n"
+        "\n"
+        "Usage: lattice-lmrescore-kaldi-rnnlm [options] \\\n"
+        "             <embedding-file> <raw-rnnlm-rxfilename> \\\n"
+        "             <lattice-rspecifier> <lattice-wspecifier>\n"
+        " e.g.: lattice-lmrescore-kaldi-rnnlm --lm-scale=-1.0 \\\n"
+        "              word_embedding.mat \\\n"
+        "              --bos-symbol=1 --eos-symbol=2 \\\n"
+        "              final.raw ark:in.lats ark:out.lats\n";
+
+    ParseOptions po(usage);
+    rnnlm::RnnlmComputeStateComputationOptions opts;
+
+    int32 max_ngram_order = 3;
+    BaseFloat lm_scale = 1.0;
+
+    po.Register("lm-scale", &lm_scale, "Scaling factor for language model "
+                "costs");
+    po.Register("max-ngram-order", &max_ngram_order,
+        "If positive, allow RNNLM histories longer than this to be identified "
+        "with each other for rescoring purposes (an approximation that "
+        "saves time and reduces output lattice size).");
+    opts.Register(&po);
+
+    po.Read(argc, argv);
+
+    if (po.NumArgs() != 4) {
+      po.PrintUsage();
+      exit(1);
+    }
+
+    if (opts.bos_index == -1 || opts.eos_index == -1) {
+      KALDI_ERR << "You must set --bos-symbol and --eos-symbol options";
+    }
+
+    std::string lats_rspecifier, word_embedding_rxfilename,
+        word_symbols_rxfilename, rnnlm_rxfilename, lats_wspecifier;
+
+    word_embedding_rxfilename = po.GetArg(1);
+    rnnlm_rxfilename = po.GetArg(2);
+    lats_rspecifier = po.GetArg(3);
+    lats_wspecifier = po.GetArg(4);
+
+    kaldi::nnet3::Nnet rnnlm;
+    ReadKaldiObject(rnnlm_rxfilename, &rnnlm);
+
+    KALDI_ASSERT(IsSimpleNnet(rnnlm));
+
+    CuMatrix<BaseFloat> word_embedding_mat;
+    ReadKaldiObject(word_embedding_rxfilename, &word_embedding_mat);
+
+    const rnnlm::RnnlmComputeStateInfo info(opts, rnnlm, word_embedding_mat);
+
+    // Reads and writes as compact lattice.
+    SequentialCompactLatticeReader compact_lattice_reader(lats_rspecifier);
+    CompactLatticeWriter compact_lattice_writer(lats_wspecifier);
+
+    int32 n_done = 0, n_fail = 0;
+
+    rnnlm::KaldiRnnlmDeterministicFst rnnlm_fst(max_ngram_order, info);
+
+    for (; !compact_lattice_reader.Done(); compact_lattice_reader.Next()) {
+      std::string key = compact_lattice_reader.Key();
+      CompactLattice clat = compact_lattice_reader.Value();
```

Comment:
Due to changes in how the reader classes are implemented, you can now make 'clat' a reference, and delete the line that says FreeCurrent().  You can do the same in wherever you got this code from, as well.